python/src/boombox/endpoints.py

```diff
@@ -0,0 +1,395 @@
+
+"""
+
+from dataclasses import dataclass, KW_ONLY, fields, is_dataclass
+from term import Atom
+from typing import Any, Literal
+from typing_extensions import override
+from abc import ABC
+
+
+@dataclass
+class BoomboxEndpoint(ABC):
+    """Abstract base class of a Boombox endpoint.
+
+    Boombox endpoints are the definitions of Boombox inputs and outputs. This
+    class is a base class for these definitions. When creating a Boombox
+    instance it expects a specification of it's input and output as
+    BoomboxEndpoints.
+
+    Attributes
+    ----------
+    transcoding_policy : {None, "if_needed", "always", "never"}
+        Allowed only for output. The default transcoding behavior is "if_needed",
+        which means that if the format of the media is the same for input and
+        output, then the stream is not decoded and encoded. This approach saves
+        resources and time, but in some cases transcoding can be necessary. To
+        force transcoding regardless if the formats differ or not, this option
+        can be set to "always". If set to "never", Boombox will never transcode,
+        raising if the desired operation can't be done without transcoding.
+    """
+
+    _: KW_ONLY
+    transcoding_policy: Literal["if_needed", "always", "never"] | None = None
+
+    def get_atom_fields(self) -> set[str]:
+        """:meta private:"""
+        return {"transcoding_policy"}
+
+    # TODO: consider checking whether an endpoint with given attributes is
+    #  valid for a direction, like so:
+    # def validate_direction(self, direction: Literal['input', 'output']) ->
+    # bool: ...
+
+    def serialize(self) -> tuple:
+        """Serializes itself to an Elixir-compatible term.
+
+        To allow Pyrlang to send the endpoint definition to Elixir it first
+        needs to be serialized into an Elixir-compatible term. This function
+        serializes the endpoint to a tuple that matches the structure of
+        Boombox endpoints in Elixir.
+
+        Returns
+        -------
+        endpoint_tuple : tuple
+            A tuple representing the endpoint. The first element is an Atom
+            representing the endpoint name, next elements are required field
+            values and in case there are any keyword fields they are in the
+            last element.
+        """
+
+        assert is_dataclass(self)
+        atom_fields = self.get_atom_fields()
+        required_field_values = [
+            self._serialize_field(f.name, atom_fields)
+            for f in fields(self)
+            if not f.kw_only
+        ]
+        keyword_fields = [
+            (Atom(f.name), self._serialize_field(f.name, atom_fields))
+            for f in fields(self)
+            if f.kw_only and self.__dict__[f.name] is not None
+        ]
+        if not keyword_fields:
+            return (self.get_endpoint_name(), *required_field_values)
+        else:
+            return (self.get_endpoint_name(), *required_field_values, keyword_fields)
```

Comment:
[NIT]

```suggestion
        if keyword_fields:
            return (self.get_endpoint_name(), *required_field_values, keyword_fields)
        else:
            return (self.get_endpoint_name(), *required_field_values)
```

(that's a shame there is no `unless` keyword in Python :<)